docs: add ztvp-certificates scenario documentation (#130)
validatedpatterns/layered-zero-trust: docs/ztvp-certificates.md

## Extraction Phases

The extraction script runs through a deterministic sequence of phases. Each
phase is independently gated by values, so the script adapts to the active
configuration.

| Phase | Gate | What It Does |
|---|---|---|
| 1 -- Custom CA | `customCA.secretRef.enabled` | Reads a user-supplied secret and writes `custom-ca.crt` |
| 2 -- Ingress CA | `autoDetect` | Loops over every `IngressController`, extracts `tls.crt` from the referenced or default router secret |
| 3 -- Service CA | `autoDetect` | Reads `openshift-service-ca.crt` ConfigMap |
| 4 -- Cluster CA Bundle | `autoDetect` | Reads `trusted-ca-bundle` from `openshift-config-managed` (present when a corporate proxy injects CAs) |
| 5 -- Additional Certs | `customCA.additionalCertificates[]` | Reads each additional secret and writes a `.crt` file |
| 6 -- Validation | `validation.enabled` | Checks minimum size and `openssl x509` parse for every `.crt` |
| 7 -- Combine | always | Concatenates all `.crt` files into `tls-ca-bundle.pem`; fails if bundle < 100 bytes |
| 8 -- ConfigMap | always | `oc apply` the `ztvp-trusted-ca` ConfigMap with annotations recording extraction metadata |

> | 8.5 -- Proxy CA | `proxyCA.enabled` | Creates a separate ConfigMap with ingress + service CAs only |

Reviewer: Any reason this is titled 8.5/8.6 instead of 8.1/8.2?

Author: updated with 8.1/8.2

> | 9 -- Image Pull Trust | `imagePullTrust.enabled` | Creates a ConfigMap keyed by registry hostname and patches `image.config.openshift.io/cluster` |

Reviewer 2: ```suggestion
| 9 -- Image Pull Trust | `imagePullTrust.enabled` | Creates a ConfigMap keyed by registry hostname and patches `image.config.openshift.io/cluster` to set the `additionalTrustedCA` attribute. |
```

Author: updated

   each registry hostname as a key and the ingress CA PEM as the value.
3. `image.config.openshift.io/cluster` is patched to set
   `additionalTrustedCA.name` to that ConfigMap.
4. The Machine Config Operator rolls the trust configuration out to all nodes.

### Scenario 6: Custom Source Locations

In non-standard environments where the ingress CA or service CA are stored in
different locations, `customSource` overrides the default auto-detection
targets.

```yaml
customSource:
  ingressCA:
    secretName: my-ingress-ca
    secretNamespace: my-namespace
    secretKey: tls.crt
  serviceCA:
    configMapName: my-service-ca
    configMapNamespace: my-namespace
    configMapKey: service-ca.crt
```

Auto-detection will read from the specified locations instead of the standard
OpenShift defaults.

## Distribution

Certificate distribution uses **ACM Governance Policies** to replicate the
`ztvp-trusted-ca` ConfigMap from `openshift-config` into each target
namespace.

```text
openshift-config/ztvp-trusted-ca  ---ACM Policy--->  qtodo/ztvp-trusted-ca
                                                      rhtpa/ztvp-trusted-ca
                                                      ...
```

> The policy uses `fromConfigMap` hub templates so that the ConfigMap data is

Reviewer: Might want to link to the ACM template documentation for `fromConfigMap`

Author: link added

always sourced from the hub cluster's copy. Target namespaces are configured
via `distribution.targetNamespaces`.

**Requirements:**

* ACM (Advanced Cluster Management) must be installed
* A `ManagedClusterSetBinding` for the `default` cluster set is created
  automatically by the chart
* The `Placement` targets clusters with `local-cluster: "true"`

## Automatic Rollout

When certificates are updated, consuming workloads need to pick up the new
bundle. The chart supports three rollout strategies:

| Strategy | Behavior |
|---|---|
| `labeled` (default) | Restarts Deployments/StatefulSets matching `ztvp.io/uses-certificates: "true"` in distribution target namespaces |
| `all` | Restarts all Deployments/StatefulSets in target namespaces |
| `specific` | Restarts only the named resources listed in `rollout.targets` |

To opt a workload into automatic restart, add the label:

```yaml
metadata:
  labels:
    ztvp.io/uses-certificates: "true"
```

## Sync Wave Ordering

The chart's resources are ordered within the ArgoCD sync:

| Wave | Resources |
|---|---|

> | 22 | ServiceAccount, RBAC (Role, RoleBinding, ClusterRole, ClusterRoleBinding) |

Reviewer: Maybe just include the ordering instead of the wave

Author: updated with relative ordering instead of absolute sync wave value

# ZTVP Certificates

> The `ztvp-certificates` chart manages CA certificate extraction, validation,

Reviewer: Add a link to the chart

Author: updated with link

bundling, and distribution across the Zero Trust Validated Pattern. It runs as

> an ArgoCD-managed application in the `openshift-config` namespace at sync-wave

Reviewer: ```suggestion
an Argo CD-managed application in the `openshift-config` namespace at sync-wave
```

Author: rephrased with correct spelling

> **21**, ensuring certificates are available before any workload that needs TLS

Reviewer: I'd avoid specifying the sync wave itself in case the number changes in the future

Author: removed sync-wave value

## Architecture

```text
 IngressControllers          Service CA           Cluster trusted-ca-bundle
 (openshift-ingress)      (openshift-config)     (openshift-config-managed)
        |                       |                          |
        +-----------+-----------+--------------------------+
                    |
         extract-certificates.sh    <-- runs as Job (initial) + CronJob (daily)
                    |
          validates & combines
                    |
                    v
        ConfigMap: ztvp-trusted-ca
           (openshift-config)
                    |
        +-----------+-----------+-----------------------+
        |                       |                       |
  ACM Policy distributes     proxyCA patches      imagePullTrust
  to target namespaces       proxy/cluster        patches image.config
   (e.g. qtodo)              (all platforms)      (when enabled)
```

### Kubernetes Resources

| Resource | Purpose |
|---|---|
| **ServiceAccount / RBAC** | Grants the extraction Job read access to secrets, configmaps, ingresscontrollers, and proxy across namespaces |
| **ConfigMap (script)** | Holds the templated `extract-certificates.sh` script |

> | **Job (initial)** | Runs once at first sync (sync-wave 23, `Prune=false`) to populate the CA bundle |

Reviewer: ```suggestion
| **Job (initial)** | Runs once at first sync to populate the CA bundle |
```

Author: updated without syncwave value

| 8 -- ConfigMap | always | `oc apply` the `ztvp-trusted-ca` ConfigMap with annotations recording extraction metadata |
| 8.5 -- Proxy CA | `proxyCA.enabled` | Creates a separate ConfigMap with ingress + service CAs only |
| 8.6 -- Proxy Patch | `proxyCA.enabled` | Patches `proxy/cluster` to set `trustedCA` (only if not already set to another value) |
| 9 -- Image Pull Trust | `imagePullTrust.enabled` | Creates a ConfigMap keyed by registry hostname and patches `image.config.openshift.io/cluster` |
| 10 -- Rollout | `rollout.enabled` | Restarts Deployments/StatefulSets that consume the certificate bundle |

## Scenario Handling

### Scenario 1: Cloud Cluster with Public Certificates (Default)

Applies to AWS, Azure, GCP, and any cluster whose ingress uses certificates
signed by a public CA.

**Active settings:**

* `autoDetect: true`
* `proxyCA.enabled: true` (default -- ensures ACS Central and other workloads
  that verify TLS on routes can trust the ingress CA without per-pod volume mounts)
* `imagePullTrust.enabled: false`

**What happens:**

1. The Job auto-detects the ingress CA from each `IngressController`'s router
   secret in `openshift-ingress`.

> 2. The service CA is read from `openshift-service-ca.crt`.

Reviewer: ```suggestion
2. The service CA is read from `openshift-service-ca.crt` from within the Job Pod.
```

Author: updated

3. If a cluster-wide proxy bundle exists, it is included.

> 4. All certificates are combined into `ztvp-trusted-ca` and distributed via

Reviewer: ```suggestion
4. All certificates are combined into `ztvp-trusted-ca` ConfigMap and distributed via
```

Author: updated

   ACM Policy to target namespaces.
5. A proxy CA ConfigMap (`ztvp-proxy-ca`) is created with ingress + service

>    CAs and `proxy/cluster` is patched so the Cluster Network Operator injects

Reviewer: ```suggestion
   CAs and the `proxy/cluster` is patched so the Cluster Network Operator injects
```

Author: updated
